src/compose/ComposeBox.js

```diff
@@ -363,7 +377,7 @@ class ComposeBox extends PureComponent<Props, State> {
 
   getDestinationNarrow = (): Narrow => {
     const { narrow } = this.props;
-    if (isStreamNarrow(narrow)) {
+    if (isStreamNarrow(narrow) || isTopicNarrow(narrow)) {
       const streamName = streamNameOfNarrow(narrow);
       const topic = this.state.topic.trim();
       return topicNarrow(streamName, topic || '(no topic)');
```

Comment:
Does this do the right thing in the case where you're in a topic narrow and simply composing a new message, rather than editing? We don't have a topic input then, so it's not obvious that we would have a `state.topic` and have the right value there.

It looks like the answer is that it doesn't, as of the commit where this change is made. But then a later change adds more logic to the `state.topic` initializer so that it does.

Probably the easiest fix is to pull that logic out to its own small prep commit: a commit that makes `state.topic` always have the topic we're going to send to (if we'd send a stream message at all), which will simplify several changes we want to make in this branch.

Reply:
> It looks like the answer is that it doesn't, as of the commit where this change is made. But then a later change adds more logic to the `state.topic` initializer so that it does.

I don't think this is true? In this commit, `state.topic` is set like so:

```javascript
    topic: this.props.lastMessageTopic,
```

Which we get here:

```javascript
    lastMessageTopic: getLastMessageTopic(state, props.narrow),
```

So I think `state.topic` should come from whatever's the last message is in `props.narrow` — in a topic narrow, that should be the correct topic.

Am I missing something here?

Reply:
Hmm, I guess that gets it closer! I hadn't made that connection.

I think there's still buggy behavior at least latently, in that if you're looking at a topic narrow and it's *empty*, `getLastMessageTopic` will return `''` instead of the topic from the narrow. I say "latently" because I'm not sure there's a way to navigate to a narrow with no messages in it...

Ah, no, actually I think that `''` case will happen all the time. The thing is that because this is the state initializer (in the constructor), it's getting run only when the `ComposeBox` instance is first created, and doesn't reflect updates to the props. And any time you visit a given narrow for the first time, we'll have nothing in `state.narrows` at that point.

Anyway. If some chain of reasoning like that does end up finding that the behavior works out correctly, it's definitely the sort of thing I'd want spelled out explicitly :slightly_smiling_face: (And then if it weren't already going away later in the branch, would want to try to find some refactor so that if possible we didn't rely on that kind of non-local connection in steady state.)